src/libraries/System.Linq/src/System/Linq/Max.cs

```diff
@@ -287,148 +154,57 @@ private static long Max(ReadOnlySpan<long> span)
             return value;
         }
 
-        public static double Max(this IEnumerable<double> source)
-        {
-            if (source.TryGetSpan(out ReadOnlySpan<double> span))
-            {
-                return Max(span);
-            }
-
-            double value;
-            using (IEnumerator<double> e = source.GetEnumerator())
-            {
-                if (!e.MoveNext())
-                {
-                    ThrowHelper.ThrowNoElementsException();
-                }
-
-                value = e.Current;
-
-                // As described in a comment on Min(this IEnumerable<double>) NaN is ordered
-                // less than all other values. We need to do explicit checks to ensure this, but
-                // once we've found a value that is not NaN we need no longer worry about it,
-                // so first loop until such a value is found (or not, as the case may be).
-                while (double.IsNaN(value))
-                {
-                    if (!e.MoveNext())
-                    {
-                        return value;
-                    }
+        public static double Max(this IEnumerable<double> source) => MaxFloat(source);
 
-                    value = e.Current;
-                }
+        public static double? Max(this IEnumerable<double?> source) => MaxFloat(source);
 
-                while (e.MoveNext())
-                {
-                    double x = e.Current;
-                    if (x > value)
-                    {
-                        value = x;
-                    }
-                }
-            }
+        public static float Max(this IEnumerable<float> source) => MaxFloat(source);
 
-            return value;
-        }
+        public static float? Max(this IEnumerable<float?> source) => MaxFloat(source);
 
-        private static double Max(ReadOnlySpan<double> span)
+        private static T MaxFloat<T>(this IEnumerable<T> source) where T : struct, IFloatingPointIeee754<T>
```

Comment:
It's a bit unfortunate that we need both one for `float` and one for `integer` given `Vector<T>` has a `Max`/`Min` functions and `INumber<T>` likewise has them.

Is there some special handling for `NaN` that LINQ is doing that prevents a single implementation?

Reply:
Looks like it might be treating `NaN` as not-preferred where-as `Max` propagates it instead.

We have `IFloatingPointIeee754<T>.MaxNumber` which does propagate the number instead, so maybe it'd be worth moving it down to `INumber<T>` so devs can do the right thing here...

I'd suspect the NaN pre-loop probably isn't that useful in practice as I'd expect NaNs are actually decently rare; which would allow the impls to then be the same.

Reply:
Then `decimal` could participate as well

Reply:
I opened https://github.com/dotnet/runtime/issues/68191.